src/meta/src/manager/catalog/mod.rs

```diff
@@ -358,44 +350,64 @@ where
         }
     }
 
-    /// Each connection is identified by a unique name
     pub async fn create_connection(
         &self,
         connection: Connection,
     ) -> MetaResult<NotificationVersion> {
-        let core = &mut self.core.lock().await.connection;
-        core.check_connection_duplicated(&connection.name)?;
+        let core = &mut self.core.lock().await;
+        let database_core = &mut core.database;
+        database_core.ensure_database_id(connection.database_id)?;
+        database_core.ensure_schema_id(connection.schema_id)?;
+
+        let key = (
+            connection.database_id,
+            connection.schema_id,
+            connection.name.clone(),
+        );
+        database_core.check_connection_name_duplicated(&key)?;
 
         let conn_id = connection.id;
-        let conn_name = connection.name.clone();
-        let mut connections = BTreeMapTransaction::new(&mut core.connections);
+        let mut connections = BTreeMapTransaction::new(&mut database_core.connections);
         connections.insert(conn_id, connection.to_owned());
         commit_meta!(self, connections)?;
 
-        core.connection_by_name.insert(conn_name, conn_id);
-
         let version = self
             .notify_frontend(Operation::Add, Info::Connection(connection))
             .await;
         Ok(version)
     }
 
-    pub async fn drop_connection(&self, conn_name: &str) -> MetaResult<NotificationVersion> {
-        let core = &mut self.core.lock().await.connection;
-
-        let conn_id = core
-            .connection_by_name
-            .remove(conn_name)
-            .ok_or_else(|| anyhow!("connection {} not found", conn_name))?;
+    pub async fn drop_connection(&self, conn_id: ConnectionId) -> MetaResult<NotificationVersion> {
```

Comment:
Oh, we should also drop all connections under a database when receiving `drop database` command. See `drop_database` for more details.